liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+        let uid = uuid::Uuid::new_v4();
+        let random_fragment_path = Self::temp_dir_path(&self.vault_dir_path).join(uid.to_string());
+        let mut fragment_file = File::create(&random_fragment_path).await?;
+        let mut hasher = blake3::Hasher::new();
+        let mut fragment_size = 0;
+
+        while let Some(bytes) = data.next().await {
+            let bytes = bytes?;
+            hasher.update(&bytes);
+            fragment_file.write(&bytes).await?;
+            fragment_size += bytes.len();
+        }
+
+        let key_bytes = hasher.finalize().as_bytes().to_vec();
+        let fragment_key = Key::try_from(key_bytes.clone())?;
+
+        // Verify integrity if key was provided
+        if let Some(key) = key {
+            if key != fragment_key {
+                bail!(
+                    "Fragment integrity check failed, expected key to be {key}, was {fragment_key}"
+                );
+            }
+        }
+
+        let key_string = bs58::encode(&key_bytes).into_string();
+        let valid_fragment_path = Self::fragment_dir_path(&self.vault_dir_path).join(key_string);
+        tokio::fs::rename(random_fragment_path, &valid_fragment_path).await?;
+
+        let fragment_info = FragmentInfo::new(
+            fragment_key.clone(),
+            &valid_fragment_path,
+            fragment_size as u64,
+        );
+        self.store_fragment_info(fragment_info).await?;
+
+        Ok(fragment_key)
+    }
+
+    async fn ensure_dirs(vault_dir_path: &Path) -> Result<()> {
+        debug!(
+            path = vault_dir_path.display().to_string(),
+            "ensuring vault dir"
+        );
+        tokio::fs::create_dir_all(vault_dir_path).await?;
+
+        let fragment_dir_path = Self::fragment_dir_path(vault_dir_path);
+        debug!(
+            path = fragment_dir_path.display().to_string(),
+            "ensuring fragment dir"
+        );
+        tokio::fs::create_dir(fragment_dir_path).await?;
+
+        let temp_dir_path = Self::temp_dir_path(vault_dir_path);
+        debug!(
+            path = temp_dir_path.display().to_string(),
+            "ensuring temp dir"
+        );
+        tokio::fs::create_dir(temp_dir_path).await?;
+
+        Ok(())
+    }
+
+    fn fragment_dir_path(vault_dir_path: &Path) -> PathBuf {
+        vault_dir_path.join(Self::FRAGMENT_DIR_NAME)
+    }
+
+    fn temp_dir_path(vault_dir_path: &Path) -> PathBuf {
+        vault_dir_path.join(Self::TEMP_DIR_NAME)
+    }
+
+    fn default_db_path(base_path: &Path) -> PathBuf {
+        base_path.join(Self::DEFAULT_VAULT_DATABASE_NAME)
+    }
+}
+
+impl Actor for Vault {
+    type Mailbox = BoundedMailbox<Self>;
+
+    async fn on_start(
+        &mut self,
+        _: kameo::actor::ActorRef<Self>,
+    ) -> std::result::Result<(), kameo::error::BoxError> {
+        self.prepare_db().await?;
+
+        Ok(())
+    }
+}
+
+impl Message<StoreFragment> for Vault {
```

Comment:
Why don;t we use the macros to create the messages automatically, like in some other places in our code? The two methods have their pros and cons. We should decide on one to limit confusion

Reply:
Sure, I will do that.

Reply:
Done, but only for store message, load requires returning impl Future + Send which does not work with this macro.